src/mpi/comm/comm_split_type.c

```diff
@@ -173,8 +173,22 @@ static int node_split_processor(MPIR_Comm * comm_ptr, int key, const char *hintv
     if (obj_containing_cpuset->type == query_obj_type) {
         color = obj_containing_cpuset->logical_index;
     } else {
-        hwloc_obj_t hobj = hwloc_get_ancestor_obj_by_type(MPIR_Process.topology, query_obj_type,
-                                                          obj_containing_cpuset);
+        hwloc_obj_t hobj = NULL;
+        if (query_obj_type == HWLOC_OBJ_NUMANODE) {
+            hwloc_obj_t tmp = NULL;
+            while ((tmp =
+                    hwloc_get_next_obj_by_type(MPIR_Process.topology, query_obj_type,
+                                               tmp)) != NULL) {
+                if (hwloc_bitmap_isincluded(obj_containing_cpuset->cpuset, tmp->cpuset) ||
+                    hwloc_bitmap_isequal(tmp->cpuset, obj_containing_cpuset->cpuset)) {
+                    hobj = tmp;
+                    break;
+                }
+            }
+        } else {
+            hobj = hwloc_get_ancestor_obj_by_type(MPIR_Process.topology, query_obj_type,
+                                                  obj_containing_cpuset);
```

Comment:
Why is this needed?  Why can't the above logic in the `if` branch work also for non-NUMANODE objects too?

Reply:
The same logic above works, I'll remove the else block.

Reply:
I have updated the patch with the else block removed and tested the changes too.